include performance improvements for FastTrack2 (memory + run-time)

taranbis opened the pull request:
• reduced memory consumption on the VarState class
	• improved parallelism on the VarState hash map by removing the Fasttrack lock and introduced locking at sub map level
	• introduced 3 memory address removal policies + 1 default to cap memory consumption
	• modified removal functions to also remove elements from the hash map used for storing call stack information. This includes replacing make_shared() function in Fasttrack::create_thread()
	• changed unit tests to be more general, to work via the general interface using log counters 
	• included the implementation of the cache efficient prefix tree (not currently used, only experimental, but maybe will useful later after some further optimizations

Commit:
perf: improve memory consumption by reducing VarState class size

- removed pointer for read-shared state from VarState class and replaced
  it with a hash map mapping memory addresses to the respective vector
  clocks in case of a read-shared state
- reduce read and write epoch of the VarState class from 64 bits
  (32 bits thread ID + 32 clock value) to 32 bits (11 bits thread num +
  21 bits clock value)
- remove size member of the VarState class

Result: sizeof(VarState) = 8

- increased concurrency by using internal locks of the parallel hash map
  used to store VarStates
- switched std::list to std::deque for callstack information for
  improved performance
- FastTrack tests now work via the general interface (easier to extend)
- Functions return_stack_trace() and set_read_write() moved to
  ThreadState as by doing so it provides better decoupling with the
  callstack storage implementation, as storing the call stack will
  just have to implement 4 functions regardless of the implementation.
  These functions are accesses from Fasttrack class or from ThreadState

Commit:
perf: add implementation for storing call stacks as a prefix tree

- Implementations for storing call stack information will just have to
  implement 4 functions: make_trace(), get_current_element(),
  insert_function_element() and remove_function_element();
- PrefixTree_StackDepot is the first version of adapting a prefix tree
  to be used for storing call stack information. It uses hash maps
  instead of constant size arrays;
- PrefixTreeDepot.h represents the cache efficient optimized version for
  the prefix tree implementation.

taranbis commented:
Hi, 

I solved the conflicts as well, as mainly all them were because a context was added to detector initialization.

fmoessbauer commented:
Hi @taranbis , thanks for this contribution.
That is a lot of code to review, so it will take some time for us to review.

May I ask you to try to cleanup the commit history a bit in the meantime.
There seems to be a lot of "working-commit" commits.
Maybe you could group that up into features and try one commit per feature.

As you wrote, the effectiveness of some optimizations depends strongly on the application under test.
Currently we do not have a clean interface to pass arguments into the detector (we have to register all options on DRace currently).

@mhmdkanj Could you take a look into that and check what is possible with `clipp` to do so?
I'd prefer a method on the detector where it registers all options along with a docstring.
Also, we could think of a "detector-options" group, which is first constructed when the detector is known (might be a bit tricky because selecting the detector itself is already an option).
Anyways, maybe we should discuss this in a separate issue.

Commit:
feat: add removal policies for capping memory consumption

- this is not an improvment feature, but a necessary one, especially
  useful in the case of long-running applications
- 4 types of removal policies, 1 default removing retired memory
  addresses and 3 that can be selected: random removal, hash map pruning
  and removal by lowest clock
- VarStates can be removed even if they are in a read-shared state

Commit:
feat: add a benchmark to measure different performance aspects

- benchmark can be modified to test different performance aspects of
  DRace, such as more memory accessed and less threads or more threads
  and less memory accessed

taranbis commented:
Hi @fmoessbauer , 

I've arranged the commit history. will look over it again, as there were many commits. The project compiles and runs flawlessly in any of these commits. 

As a TODO for me is still to check if there is a run-time drawback in using a stack of thread numbers instead of linearly increasing thread numbers. If there's not I will change the first commit to work with a stack of thread numbers, otherwise we will discuss here.

mhmdkanj commented:
Hi @taranbis ,

Please check the latest `master` as from now on you could pass detector-specific options using the CLI, and if you're using the GUI, there as well. In brief, you could do that after specifying the detector flag with `-d fasttrack -option1 -option2` - you could see that in the DRace README for the usage.

In case you had questions/concerns/bugs with this usage please feel free to contact me!

mhmdkanj commented on standalone/detectors/fasttrack/include/fasttrack.h:
```diff
@@ -308,326 +705,389 @@ class Fasttrack : public Detector {
+      vars.dropSubMap((index & mask));
+      for (auto it = threads.begin(); it != threads.end(); ++it) {
+        // as there is no randomness same SubMap is dropped.
+        it->second->read_write.dropSubMap((index & mask));
```
Please double-check compilation here:

On my machine, lines with `x.dropSubMap()` produced a compiler error: 
```
'dropSubMap': is not a member of 'phmap::parallel_node_hash_map<size_t,VarState,phmap::Hash<uint64_t>,phmap::EqualTo<T>,std::allocator<std::pair<const K,V>>,5,ipc::spinlock>'
```

mhmdkanj commented on standalone/detectors/fasttrack/include/fasttrack.h:
```diff
@@ -1,40 +1,50 @@
 #ifndef FASTTRACK_H
 #define FASTTRACK_H
+#pragma once
```
Please remove redundant `pragma` since include guards already exist and is used consistently.

mhmdkanj commented on standalone/detectors/fasttrack/include/fasttrack.h:
```diff
@@ -308,326 +705,389 @@ class Fasttrack : public Detector {
 }  // namespace drace
-
-#endif  // !FASTTRACK_H
+#endif  // !FASTTRACK_H
```
Please amend: GitHub complains that `No newline at end of file`

mhmdkanj commented on standalone/detectors/fasttrack/include/fasttrack.h:
```diff
@@ -150,124 +468,197 @@ class Fasttrack : public Detector {
 
-    if (v->is_read_shared() &&
-        v->get_vc_by_thr(tid) == id) {  // read shared same epoch
+    if (shared_vc != nullptr && v->get_vc_by_th_num(th_num, shared_vc) ==
```
Please swap the `nullptr` in `x == nullptr` and `x != nullptr` to be on the left-hand side (e.g. `nullptr != x`) of the comparison operator to avoid potential overloaded operators. 
Also in other lines at least in 524, 553, 562, 624

mhmdkanj commented on standalone/detectors/fasttrack/include/MemoryPool.h:
```diff
@@ -0,0 +1,61 @@
+  MemoryPool() = default;
+  MemoryPool(size_t chunkSize) : chunk_size(chunkSize) {}
+  MemoryPool(size_t chunkSize, size_t numChunks)
+      : chunk_size(chunkSize), num_chunks(numChunks) {
```
Please stick to snake case for variables such as constructor parameters, just as `num_chunks` or `num_chunks_value`

mhmdkanj commented on standalone/detectors/fasttrack/include/MemoryPool.h:
```diff
@@ -0,0 +1,61 @@
+  size_t chunk_size;
+
+  /// block size
+  size_t block_size;
```
If applicable, please initialize these private members (e.g. to `0`)

mhmdkanj commented on standalone/detectors/fasttrack/include/MemoryPool.h:
```diff
@@ -0,0 +1,61 @@
+  void print_used_memory() {
+    std::cout << "Memory Allocated: " << chunks_allocated * chunk_size
+              << std::endl;
+  }
```
Possibly move this function to the implementation file.

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+ *   Felix Moessbauer <felix.moessbauer@siemens.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
```
Execute clang style formatting script on this file for it to pass the pipeline.

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+  }
+};
+
+#endif  //! POOL_ALLOCATOR_HEADER_H
```
Please amend: GitHub complains that `No newline at end of file`

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+  size_type max_size() const { return std::numeric_limits<size_type>::max(); }
+
+  pointer allocate() {
+    if (free_pointer.load(std::memory_order_release) == nullptr) {
```
Please swap the `nullptr` in `x == nullptr` and `x != nullptr` to be on the left-hand side (e.g. `nullptr != x`) of the comparison operator to avoid potential overloaded operators.

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+ *------------------------------------------------------------------------------
+ *
+ * Header File implementing a pool allocator. It can also be used
+ * in combination with the Segregator class, such as on line 136.
```
Try avoiding to refer to a specific line in this comment-documentation, but rather reference it with a doxy command or just state the var name and reference it

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+  };
+};
+
+template <typename T, size_t numChunks = 512>
```
Please stick to snake case for the templ variable `numChunks` -> `num_chunks` or similar.
Also occurs at least in lines 109, 130, 144,..

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+};
+
+template <typename T, size_t numChunks = 512>
+class PoolAllocator {
```
Try briefly investigating whether the classes in this file could be segregated into several files, and possibly avoid code duplication (as to member functions) among them with inheritance.

taranbis replied:
Maybe only an interface, as all of them have a static member, that is the memory pool and inheritance would certainly not help code duplication but it will offer a proper framework.

mhmdkanj replied:
@fmoessbauer

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+MemoryPool SizePoolAllocator<size, numChunks>::mem_pool(size, numChunks);
+
+template <class T>
+using DRaceAllocator = Segregator<
```
Maybe add some documentation above this to elaborate on the (possible?) magic numbers in the template.

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+  using pointer = T*;
+  using const_pointer = const T*;
+  using void_pointer = void*;
+  using const_void_pointer = const void*;
```
Are all these simplifications used in the classes and other files? In case some are not necessary maybe remove those.

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,278 @@
+
```
Remove empty first line

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,278 @@
+  }
+};
+
+#endif
```
Comment here header guard
```suggestion
#endif // TREEDEPOT_HEADER_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,278 @@
+  void insert_function_element(size_t pc) {
+    std::lock_guard<ipc::spinlock> lg(read_write_lock);
+
+    if (_curr_elem == nullptr) {
```
Please swap the `nullptr` in `x == nullptr` and `x != nullptr` to be on the left-hand side (e.g. `nullptr != x`) of the comparison operator to avoid potential overloaded operators. 
Also at least in lines: 252, 254, 270,..

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,278 @@
+ *------------------------------------------------------------------------------
+ */
+
+class INode {
```
Assuming `INode` is an interface class: rather than throwing an exception in the implemented functions, would it work to make them abstract and not define an implementation for this class?

taranbis replied:
Hi, 
It is done like this, because when a node is created, it is first created through the abstract class and then specialized on the needs, hence I couldn't do with an abstract class and I had to find a way to not allow any client use the base function.

mhmdkanj replied:
@fmoessbauer

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,278 @@
+ *   Felix Moessbauer <felix.moessbauer@siemens.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
```
Is this file supposed to function like a library for it to include several classes? Otherwise, maybe investigate the possibility of separating it.

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,278 @@
+    }
+  }
+
+  size_t size() const final { return N; }
```
Please change this getter's name to indicate that it is a function with a verb in the name, just like `get_size()` for instance.

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTree_StackDepot.h:
```diff
@@ -0,0 +1,95 @@
+ public:
+  TrieNode* get_current_element() { return _curr_elem; }
+
+  void InsertFunction(size_t pc) {
```
Please stick to snake case convention like other function names to be consistent, for e.g. `insert_function()`

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTree_StackDepot.h:
```diff
@@ -0,0 +1,95 @@
+    _curr_elem->pc = pc;
+  }
+
+  void ExitFunction() {
```
Please stick to snake case convention like other function names to be consistent, for e.g. `exit_function()`

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTree_StackDepot.h:
```diff
@@ -0,0 +1,95 @@
+  TrieNode* get_current_element() { return _curr_elem; }
+
+  void InsertFunction(size_t pc) {
+    if (_curr_elem == nullptr) {
```
Please swap the `nullptr` in `x == nullptr` and `x != nullptr` to be on the left-hand side (e.g. `nullptr != x`) of the comparison operator to avoid potential overloaded operators. 
Also at least in line 88,..

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTree_StackDepot.h:
```diff
@@ -0,0 +1,95 @@
+    return std::move(this_stack);
+  }
+};
+#endif
```
Comment next to header guard
```suggestion
#endif // PREFIXTREE_STACKDEPOT_HEADER_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/stacktrace.h:
```diff
@@ -29,63 +32,49 @@ class StackTrace {
+  StackTree::vertex_descriptor get_current_element() const;
 };
-#endif
+#endif
```
Please amend: GitHub complains that `No newline at end of file`
In addition, comment next to header guard:
```suggestion
#endif // _STACKTRACE_H_
```

mhmdkanj commented on standalone/detectors/fasttrack/include/threadstate.h:
```diff
@@ -36,21 +84,40 @@ class ThreadState : public VectorClock<> {
+   */
+  std::deque<size_t> return_stack_trace(std::size_t address) const;
+};
+#endif  // !THREADSTATE_H
```
Please amend: GitHub complains that No newline at end of file

mhmdkanj commented on standalone/detectors/fasttrack/include/varstate.h:
```diff
@@ -1,15 +1,18 @@
  *
  * Authors:
  *   Felix Moessbauer <felix.moessbauer@siemens.com>
+ *   Mihai Robescu <mihai-gabriel.robescu@siemens.com>
```
Execute clang style formatting script on this file to pass the pipeline

mhmdkanj commented on standalone/detectors/fasttrack/include/varstate.h:
```diff
@@ -22,97 +25,112 @@
+      xvector<VectorClock<>::VC_EPOCH>* shared_vc) const;
 };
-#endif  // !VARSTATE_H
+#endif  // !VARSTATE_H
```
Please amend: GitHub complains that No newline at end of file

mhmdkanj commented on standalone/detectors/fasttrack/include/vectorclock.h:
```diff
@@ -102,29 +115,72 @@ class VectorClock {
+  }
+
+ private:
+  static bool _thread_no_initiliazation;
```
Initialize this member if applicable

mhmdkanj commented on standalone/detectors/fasttrack/include/vectorclock.h:
```diff
@@ -102,29 +115,72 @@ class VectorClock {
+VectorClock<>::ThreadNum VectorClock<>::thread_no = 1;
+phmap::flat_hash_map<VectorClock<>::ThreadNum, VectorClock<>::TID>
+    VectorClock<>::thread_ids;
+#endif
```
Please amend: GitHub complains that No newline at end of file

mhmdkanj commented on standalone/detectors/fasttrack/src/MemoryPool.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "MemoryPool.h"
```
Add missing copyright & authors header

mhmdkanj commented on standalone/detectors/fasttrack/src/MemoryPool.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "MemoryPool.h"
+
+Chunk* MemoryPool::do_allocation() {
+  if (free_pointer == nullptr) {
```
Please swap the `nullptr` in `x == nullptr` and `x != nullptr` to be on the left-hand side (e.g. `nullptr != x`) of the comparison operator to avoid potential overloaded operators.

mhmdkanj commented on standalone/detectors/fasttrack/src/MemoryPool.cpp:
```diff
@@ -0,0 +1,31 @@
+  c->next = free_pointer;
+  free_pointer = c;
+  chunks_allocated--;
+}
```
Please amend: GitHub complains that `No newline at end of file`

mhmdkanj commented on standalone/detectors/fasttrack/src/threadstate.cpp:
```diff
@@ -1,26 +1,49 @@
+  // A read/write operation was not tracked correctly => return empty
+  // stacktrace
+  return {};
+}
```
Please amend: GitHub complains that `No newline at end of file`

mhmdkanj commented on standalone/detectors/fasttrack/src/varstate.cpp:
```diff
@@ -113,18 +87,22 @@ VectorClock<>::VC_ID VarState::get_sh_id(uint32_t pos) const {
   }
   return 0;
-}
+}
```
Please amend: GitHub complains that `No newline at end of file`

mhmdkanj commented on standalone/detectors/fasttrack/src/varstate.cpp:
```diff
@@ -12,18 +12,19 @@
 bool VarState::is_ww_race(ThreadState* t) const {
-  if (get_write_id() != VAR_NOT_INIT && t->get_tid() != get_w_tid() &&
-      get_w_clock() >= t->get_clock_by_tid(get_w_tid())) {
+  if (get_write_epoch() != VAR_NOT_INIT &&
```
Execute clang style formatting script so that the file passes the pipeline

mhmdkanj commented on standalone/ft_benchmark/ft_benchmark.cpp:
```diff
@@ -0,0 +1,146 @@
+ *   Felix Moessbauer <felix.moessbauer@siemens.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
```
Execute clang style formatting script for this file to pass the pipeline

mhmdkanj commented on standalone/ft_benchmark/ft_benchmark.cpp:
```diff
@@ -0,0 +1,146 @@
+  // std::cin.get();
+}
+
+int CountPossibleDataRaces() {
```
Try sticking to snake case naming as in `count_possible_data_races()` for e.g.

mhmdkanj commented on standalone/detectors/fasttrack/include/MemoryPool.h:
```diff
@@ -0,0 +1,61 @@
+#ifndef MEMORY_POOL_HEADER_H
+#define MEMORY_POOL_HEADER_H 1
+#pragma once
```
Please remove insignificant `1` and the redundant pragma since include guards already exist and are used 
```suggestion
#define MEMORY_POOL_HEADER_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,199 @@
+#ifndef POOL_ALLOCATOR_HEADER_H
+#define POOL_ALLOCATOR_HEADER_H 1
+#pragma once
```
Please remove insignificant `1` and the redundant pragma since include guards already exist and are used 
```suggestion
#define POOL_ALLOCATOR_HEADER_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,278 @@
+
+#ifndef TREEDEPOT_HEADER_H
+#define TREEDEPOT_HEADER_H 1
+#pragma once
```
Please remove insignificant `1` and the redundant pragma since include guards already exist and are used 
```suggestion
#define TREEDEPOT_HEADER_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTree_StackDepot.h:
```diff
@@ -0,0 +1,95 @@
+#ifndef PREFIXTREE_STACKDEPOT_HEADER_H
+#define PREFIXTREE_STACKDEPOT_HEADER_H 1
+#pragma once
```
Please remove insignificant `1` and the redundant pragma since include guards already exist and are used 
```suggestion
#define PREFIXTREE_STACKDEPOT_HEADER_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/stacktrace.h:
```diff
@@ -1,5 +1,6 @@
-#define _STACKTRACE_H_
+#ifndef STACKTRACE_H
+#define STACKTRACE_H 1
+#pragma once
```
Please remove insignificant `1` and the redundant pragma since include guards already exist and are used 
```suggestion
#define STACKTRACE_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/threadstate.h:
```diff
@@ -1,32 +1,80 @@
 #ifndef THREADSTATE_H
-#define THREADSTATE_H
+#define THREADSTATE_H 1
+#pragma once
```
Please remove insignificant `1` and the redundant pragma since include guards already exist and are used 
```suggestion
#define THREADSTATE_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/varstate.h:
```diff
@@ -1,15 +1,18 @@
+#ifndef VARSTATE_HEADER_H
+#define VARSTATE_HEADER_H 1
+#pragma once
```
Please remove insignificant `1` and the redundant pragma since include guards already exist and are used 
```suggestion
#define VARSTATE_HEADER_H
```

mhmdkanj commented on standalone/detectors/fasttrack/include/vectorclock.h:
```diff
@@ -1,5 +1,7 @@
-#define VECTORCLOCK_H
+#ifndef VECTORCLOCK_HEADER_H
+#define VECTORCLOCK_HEADER_H 1
+#pragma once
```
Please remove insignificant `1` and the redundant pragma since include guards already exist and are used 
```suggestion
#define VECTORCLOCK_HEADER_H
```

mhmdkanj commented on standalone/detectors/fasttrack/src/stacktrace.cpp:
```diff
@@ -58,65 +59,49 @@ void StackTrace::clean() {
-  // trace
-  return {};
+  return curr_elem;
 }
```
Please amend: GitHub complains that `No newline at end of file` (shows in the last commit rather than total changes)

mhmdkanj commented on standalone/detectors/fasttrack/src/threadstate.cpp:
```diff
@@ -1,26 +1,49 @@
+  auto it = read_write.find(addr);
+  if (it == read_write.end()) {
+    // TODO: maybe use std::move avoid copy on pair
+    // should I do sth if traceDepot.GetCurrentElement() is nullptr??
```
Maybe discuss this here in the comments or in the PR overview and change/don't change accordingly.

mhmdkanj commented on standalone/ft_benchmark/ft_benchmark.cpp:
```diff
@@ -0,0 +1,146 @@
+      auto tmp = *iter;
+      iter++;
+    }
+    // std::this_thread::sleep_for(std::chrono::milliseconds(500));
```
If this was only used for debugging, try removing the comment if it is not relevant anymore.
Similarly in lines 74, 125, 134.

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,278 @@
+ * cache line
+ *
+ *------------------------------------------------------------------------------
+ */
```
This file has a couple of commented out code - would be good to rethink if some of them are necessary to keep, otherwise it's better to avoid the "lava flow"

mhmdkanj reviewed (changes requested):
Hi @taranbis 

First of all, great effort on your additions and changes!

I've left some comments throughout the files:
- The most critical one being a compilation error in `fasttrack.h` (at least on my side), which needs a double-checking. Please refer to the corresponding comment for more info.
- Most of the remaining comments pertain to style changes in order to be consistent with all other files - should be fast to change.
- A couple of these comments pertain to questions regarding the intended use (in the case where several classes are put together in one header file) and whether this could be further beautified.

Upon making further changes and committing, please make sure of the following:
- The clang style formatting script is executed on the changed file
- The PR is rebased with the new master

After resolving these comments, we will further review the changes as to their implementation and usage.

Best Regards,
Mohamad

taranbis commented:
Hi @mhmdkanj,

I will look within the next days over your comments. For the moment I just want to talk about the compilation error in ```fasttrack.h```. I changed a file in the phmap submodule, namely, ```phmap.h``` to be able to drop an entire sub hash map, as I couldn't work with just the public interface. This is why the function is missing.

The changes cannot be commited from my side. They could be placed as a patch to the phmap implementation. 

Best greetings, 
Mihai

mhmdkanj commented:
> Hi @mhmdkanj,
> 
> I will look within the next days over your comments. For the moment I just want to talk about the compilation error in `fasttrack.h`. I changed a file in the phmap submodule, namely, `phmap.h` to be able to drop an entire sub hash map, as I couldn't work with just the public interface. This is why the function is missing.
> 
> The changes cannot be commited from my side. They could be placed as a patch to the phmap implementation.
> 
> Best greetings,
> Mihai

@taranbis In the course of solving this issue, try to avoid the need to fork the `phmap` repo in order to patch it - but rather come up with an implementation on the DRace side, or if `greg7mdp` can patch it by opening up an issue there, etc... This can be up for discussion as well.

taranbis commented:
Hi @mhmdkanj , 

well this feature is used to remove data from tracking and I am working with private members of the ```phmap``` class as I am dropping an entire sub hash map of memory addresses. @fmoessbauer knows about this. I don't have a solution to keep the feature will not modifying at least something in the fast parallel hash map implementation. 

1. to patch the hash map implementation.
2. it could also work if the array of sub hash map in the ```phmap``` implementation is declared public, as it is currently private. This is why I had to work create the functions within that class. If it is public, I could move all implementations to ```DRace``` side However, it would not make much of a sense to declare the array of sub hash maps as public.
3. just drop the feature, but I think it is faster than random or lowest clock removals. I could come up with some measurements if dropping this feature comes into discussion. 

of course, I am open to suggestions and the choice is yours in the end.

All the best, 
Mihai

fmoessbauer commented:
Hi @taranbis , I did already discuss this internally with @mhmdkanj and also back then was aware of the issue.
Nonetheless, we want to avoid having to maintain a fork of the phmap. By that, we have two options:

1. upstream the necessary feature. As we have a use-case for that, we could try to get into contact with greg7mdp and discuss options for integration. Maybe just declaring some members protected would help, so that we can inherit from the map.
2. Use another strategy that does not rely on the internals of the phmap. But according to your results, this was the most promising strategy, was it?

Anyways, thanks for this well-written contribution and for your interest in speeding up DRace ;)

taranbis commented:
Hi @fmoessbauer , Hi @mhmdkanj, 

I didn't get the chance to look over all your proposed changes. However, what I have done is talk to `greg7mdp` and he added the necessary API to be able to drop a sub hash map, so we can maintain all proposed features:

```cpp
    void clear(std::size_t submap_index) {
        Inner& inner = sets_[submap_index];
        typename Lockable::UniqueLock m(inner);
        inner.set_.clear();
    }
```

Within the last commit I added the new API functionality and I also changed some other things that failed due to a changed namespace name. Now it should compile fine, so you can test in the meantime while I modify also the other things

Best regards!

fmoessbauer commented:
Hi @taranbis,

> However, what I have done is talk to `greg7mdp` and he added the necessary API to be able to drop a sub hash map, so we can maintain all proposed features:

Excellent work! Now we can just use the upstream version and benefit from (possible) improvements there, instead of maintaining our own fork which would bitrot at some point in time.

Thanks also for taking the time to fix the other mentioned points. Is there still something conceptual, where we can help you? Or anything that needs a discussion?

All the best, Felix

taranbis commented:
Hi, 

I am sorry for the long due reply, but I have been really busy. First of all, I have walked though all the proposed changes and did all the required changes. However, working though the project again I realized that something could be improved, namely replacing pointers in prefix trees with shared pointers and also counting references in order to be able to remove them when needed. 

However, this might take some time for me to investigate, but I will happily do it to see that memory is ok deallocated and that there is a mechanism for future removals.

All the best, 
Mihai

fmoessbauer commented:
Hi Mihai, thanks for taking care of integrating the required changes.
Just take your time on the shared-ptr stuff.

I think we can integrate that in the meantime. Then, just create a new PR for that.

@mhmdkanj : Could you please give the new version a try?

Best, Felix

mhmdkanj commented on standalone/ft_benchmark/ft_benchmark.cpp:
```diff
@@ -0,0 +1,142 @@
+
+void read_from_block(std::vector<std::pair<uintptr_t*, uintptr_t*>>* blocks) {
+  try {
+    std::uniform_int_distribution<int> dist(0, blocks->size() - 1);
```
Line 41 generates a warning that `conversion from 'size_t' to '_Ty', possible loss of data`; if this is intended, then it would be safe to ignore. Similarly in line 61.

mhmdkanj commented on standalone/ft_benchmark/ft_benchmark.cpp:
```diff
@@ -0,0 +1,142 @@
+  } catch (const std::exception& e) {
+    std::cout << e.what() << std::endl;
+  } catch (...) {
+    std::cout << "Failure!" << std::endl;
```
Please provide here something more descriptive than "Failure!" for e.g. "Unexpected failure in module *module_name*."

mhmdkanj commented on standalone/detectors/fasttrack/include/SelectAllocator.h:
```diff
@@ -0,0 +1,70 @@
+  void deallocate(INode* ptr, size_t size) {
+    if (size < threshold1) {
+      // Node<threshold1>* tmp = dynamic_cast<Node<threshold1>*>(ptr);
+      // tmp->~Node<threshold1>(); //doens't work when I am calling destructor
```
Maybe delete if not needed, otherwise can be discussed in a separate issue/in this review.

mhmdkanj commented on standalone/detectors/fasttrack/include/StackTreeNode.h:
```diff
@@ -0,0 +1,96 @@
+  }
+};
+
+#endif  //! STACKTREE_NODE_H
```
GitHub complains that "No newline at end of file"

mhmdkanj commented on standalone/detectors/fasttrack/include/PoolAllocator.h:
```diff
@@ -0,0 +1,196 @@
+/**
+ * \class PoolAllocator.
+ * \brief it used to allocate memory in chunks. It can be used in combination
+ * with the \class Segregator to split on the sizes Header File implementing a
```
Typo: "it **is** used to allocate memory in chunks." 
Also the `\brief` portion seems to be cut off at the end with "Header File implementing a..."

mhmdkanj commented on standalone/detectors/fasttrack/include/PrefixTreeDepot.h:
```diff
@@ -0,0 +1,122 @@
+  void remove_function_element() {
+    std::lock_guard<ipc::spinlock> lg(read_write_lock);
+
+    // func_exit before func_enter might happen. The instrumentation side might skip a func_enter call
```
Please apply clang-style formatting on this file to adjust the long line so that the pipeline would succeed when checking for code style.

mhmdkanj commented on standalone/detectors/fasttrack/include/fasttrack.h:
```diff
@@ -50,66 +59,372 @@ class Fasttrack : public Detector {
+
+  /// switch logging of read/write operations
+  bool log_flag = true;
+  bool final_output = (true && log_flag);
```
When possible, rebase your branch with the new `master` so that the execution could support passing flags to `fasttrack` upon running drace, for e.g. `drrun.exe -c drace-client.dll -d fasttrack --stats --size --other-drace-flags [...]`. This feature has already been added to the `master` branch (after you branched off to make these changes). You could try it out to see if passing the flags to `fasttrack` works as you would have expected. In any case this would also take care of these merge conflicts along the way.

That way, you could revert the initial state of `log_flag` back to `false` so that the user can toggle whether or not to show the logs when a flag is passed (by default not showing them).
Consequently, any occurrences of `final_output` should be replaced with `log_flag` (especially when calling
`process_log_output();` on line 390)
```suggestion
  bool log_flag = false;
```

mhmdkanj commented on standalone/detectors/fasttrack/include/fasttrack.h:
```diff
@@ -50,66 +59,372 @@ class Fasttrack : public Detector {
+    shared_vcs.clear();
+    VectorClock<>::thread_ids.clear();
+
+    if (final_output) process_log_output();
```
Please refer to the other comment pertaining to this file.
```suggestion
    if (log_flag) process_log_output();
```

mhmdkanj reviewed (changes requested):
Hi @taranbis 

Thank you lots for incorporating these changes! The required code style is thus matched.

Also, it compiles successfully now & I tried executing the integration tests locally with `fasttrack`, whereby all had succeeded (except for the `DotnetClrMonitor` one for some reason; we can check that out later @fmoessbauer, most probably the source is elsewhere).

Please take a look at the couple of comments I posted after the new changes. Don't worry though; they are mostly code-style issues due to the new changes and/or things I forgot to review previously. So, they should be fast to incorporate. The most pressing one is rebasing the branch onto the new `master` so that logging of `fasttrack` details could be toggled on/off by the user using `fasttrack` flags rather than being toggled on by default.

Take your time with adjusting these small details, and afterwards we could move to start integrating them.
If you have any inquiries feel free to contact me.

Best Regards,
Mohamad

mhmdkanj commented:
Hi @fmoessbauer , I've mentioned you in three revision comments whereby your input might be needed, whenever you are able. Thanks in advance!

Commit:
docs: add sparsepp and parallel-hashmap to dependencies

Commit:
refactor: add descriptive logging for ft benchmark read

Commit:
docs: adjust optz ft comments

Commit:
refactor: set default logging to false in fasttrack

mhmdkanj commented:
Hi @fmoessbauer ,
Current status:
- - [x] My revisions above have already been merged
- - [x] All Integration tests pass
- - [x] 9/42 unit tests fail, all of which pertain to the newly written fasttrack tests; I can send you the log when/if needed

taranbis commented:
Hi @mhmdkanj, 

since the newly written fasttrack is responsible for that, I will have a look. I will get until Sunday back to you (sorry the last few months have been crazy). you can send the logs to my private email address.

mhmdkanj commented:
> Hi @mhmdkanj,
> 
> since the newly written fasttrack is responsible for that, I will have a look. I will get until Sunday back to you (sorry the last few months have been crazy). you can send the logs to my private email address.

Hi @taranbis  ,
I've sent you the text file to your yahoo mail.

Best,
Mohamad

taranbis commented:
Hi @mhmdkanj , 

I looked over the tests. So, the reason the tests fail is because they use the logging structured In the last commit: **refactor: set default logging to false** in fasttrack (6f7d3da75419adeeb8bea821609779672b2fd43c), you removed the _final_output_  variable and made _log_flag_ false. 

The tests used the logging structure as it was an easy way not to depend on the internals of the fasttrack algorithm and pass info to fasttrack like normally. If you undo your commit, the tests should work fine. 

I am already thinking of ways to rewrite this and one such way would be the callback. However, it doesn't hold all the info needed to differentiate the cases.  

Until then, the integration tests pass if set the log flag to true. I did also a final_output flag to not print it when it was not necessary. 

All the best, 
Mihai

taranbis commented:
Oh and one more thing. The callback not only does not hold all the required info, it is also only triggered when there's a race. You cannot distinguish the other use cases such as read-exclusive, read-shared-same-epoch etc.

Commit:
refactor: add class test wrapper

taranbis commented:
Hi @mhmdkanj, 

Nevermind. I solved it. I added a wrapper on top of fasttrack to work with the testing framework. Now logging is like in your commit (false by default) and integration tests work. See my commit **refactor: add class test wrapper** (f381862cb98164a1df9f6aa503fce43b40eb8975). 

Tell me if there is anything else. 

P.S.: try not to use force pushes :D.

All the best, 
Mihai

Commit:
style: add reuse header for fasttrack wrapper

mhmdkanj commented:
> Hi @mhmdkanj,
> 
> Nevermind. I solved it. I added a wrapper on top of fasttrack to work with the testing framework. Now logging is like in your commit (false by default) and integration tests work. See my commit **refactor: add class test wrapper** ([f381862](https://github.com/siemens/drace/commit/f381862cb98164a1df9f6aa503fce43b40eb8975)).
> 
> Tell me if there is anything else.
> 
> P.S.: try not to use force pushes :D.
> 
> All the best,
> Mihai

Hi @taranbis , 
Yup this looks good. However:
```
97% tests passed, 2 tests failed out of 59

Total Test time (real) =  21.98 sec

The following tests FAILED:
         58 - Interface/DetectorTest.Parallelism/"fasttrack.standalone" (Failed)
         59 - Interface/DetectorTest.Parallelism/"tsan" (Failed)
```
Do these two unit tests pass on your end?

taranbis commented:
Hi @mhmdkanj , 

No, but this is not  fasttrack related. it also fails for tsan. This is a test written by @fmoessbauer to check that concurrency works well on the detector side. I might be able to have a look into those, but i guess @fmoessbauer knows better. I will check nonetheless.

mhmdkanj commented:
@taranbis True, it's probably not a direct consequence but I was wondering if you were experiencing the same things.

mhmdkanj commented:
Hi @fmoessbauer ,
So on my machine, the following tests cause the missing of a 100% pass rate:

### Unit Tests
```
The following tests FAILED:
         45 - Interface/DetectorTest.ForkInitialize/"tsan" (Failed)
         47 - Interface/DetectorTest.Barrier/"tsan" (Failed)
         49 - Interface/DetectorTest.ResetRange/"tsan" (Failed)
         51 - Interface/DetectorTest.RaceInspection/"tsan" (Failed)
         58 - Interface/DetectorTest.Parallelism/"fasttrack.standalone" (Failed)
         59 - Interface/DetectorTest.Parallelism/"tsan" (Failed)

```
Most of the time only the last two fail; however, occasionally the others fail as well.

### Integration Tests
```
[  FAILED  ] Integration/DR.HowToSolution/0, where GetParam() = "fasttrack"
[  FAILED  ] Integration/DR.NoIoRaces/0, where GetParam() = "fasttrack"
```
Usually either one or both of these fail.

fmoessbauer commented:
Hi @taranbis , @mhmdkanj ,

thanks for taking care of the integration. Regarding the failed tests:

- `Interface/DetectorTest.*/"tsan"` the TSAN tests fail in case TSAN is not able to allocate memory at the specified location. That happens from time to time. Then, all TSAN tests fail. There's not much we can do about it.
- `Interface/DetectorTest.Parallelism/*` I will have a look into that one. Maybe there is an issue in the test itself.
- `Integration/DR.HowToSolution/`, `Integration/DR.NoIoRaces/0` looks like we get races on IO. Maybe symbol names changed, so they are not correctly intercepted. I will have a look.

Cheers!
